crypt.go

```diff
@@ -0,0 +1,494 @@
+// how many attempts should be made to activate the volume with the recovery key before failing. If this is set to 0, then no attempts
+// will be made to activate the encrypted volume with the fallback recovery key. If activation with the recovery key is successful,
+// the recovery key will be added to the root user keyring in the kernel with a description of the format
+// "<argv[0]>:<volumeName>:reason=<reason>" where reason is an integer that describes the recovery reason - see the RecoveryReason
+// type.
+//
+// If either the PINTries or RecoveryKeyTries fields of options are less than zero, an error will be returned. If the ActivateOptions
+// field of options contains the "tries=" option, then an error will be returned. This option cannot be used with this function.
+//
+// If the LockSealedKeyAccess field of options is true and the call to LockAccessToSealedKeys fails, a LockAccessToSealedKeysError
+// error will be returned. In this case, activation with either the TPM sealed key or the fallback recovery key will not be attempted.
+//
+// If activation with the TPM sealed key fails, a *ActivateWithTPMSealedKeyError error will be returned, even if the subsequent
+// fallback recovery activation is successful. In this case, the RecoveryErr field of the returned error will be nil, and the TPMErr
+// field will contain the original error. If activation with the fallback recovery key also fails, the RecoveryErr field of the
+// returned error will also contain details of the error encountered during recovery key activation.
+//
+// If the volume is successfully activated, either with the TPM sealed key or the fallback recovery key, this function returns true.
+// If it is not successfully activated, then this function returns false.
+func ActivateVolumeWithTPMSealedKey(tpm *TPMConnection, volumeName, sourceDevicePath, keyPath string, pinReader io.Reader, options *ActivateWithTPMSealedKeyOptions) (bool, error) {
+	if options.PINTries < 0 {
+		return false, errors.New("invalid PINTries")
+	}
+	if options.RecoveryKeyTries < 0 {
+		return false, errors.New("invalid RecoveryKeyTries")
+	}
+
+	activateOptions, err := makeActivateOptions(options.ActivateOptions)
+	if err != nil {
+		return false, err
+	}
+
+	if err := activateWithTPMKey(tpm, volumeName, sourceDevicePath, keyPath, pinReader, options.PINTries, options.LockSealedKeyAccess, activateOptions); err != nil {
+		reason := RecoveryReasonUnexpectedError
+		switch {
+		case isLockAccessError(err):
+			return false, LockAccessToSealedKeysError(err.Error())
+		case xerrors.Is(err, ErrTPMLockout):
+			reason = RecoveryReasonTPMLockout
+		case xerrors.Is(err, ErrTPMProvisioning):
+			reason = RecoveryReasonTPMProvisioningError
+		case isInvalidKeyFileError(err):
+			reason = RecoveryReasonInvalidKeyFile
+		case xerrors.Is(err, requiresPinErr):
+			reason = RecoveryReasonPINFail
```

Comment:
this doesn't seem exercised

Reply:
This requires https://github.com/snapcore/secboot/pull/34